### PR TITLE
Invoke Contract Function

### LINKED ISSUE #23: Invoke contract function
## Objective

Enable the invocation of a smart contract through a user-friendly interface.

## Checklist

- [x] Create **Contract** Interface
- [x] Add function `invoke/` to Contract interface

### PR DESCRIPTION
**Pull request by Odraxs**
## Description

Closes #23 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: lib/contract/invoke_host_function.ex [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  @type sc_val_list :: list(SCVal.t())
+
+  @spec invoke(
+          contract_id :: contract_id(),
+          source_secret_key :: source_secret_key(),
+          function_name :: function_name(),
+          function_args :: function_args(),
+          auth_accounts :: auth_accounts()
+        ) :: send_response()
+  def invoke(
```
Comment (juanhurtado10) [inline lib/contract/invoke_host_function.ex:42]: @Odraxs there is too much logic encapsulated in a single function. Try breaking it down into functional code chunks.

File: lib/contract/invoke_host_function.ex [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        function_name,
+        function_args,
+        auth_accounts \\ []
+      ) do
+    function_args = covert_to_sc_val(function_args)
+    keypair = {public_key, _secret} = Stellar.KeyPair.from_secret_seed(source_secret_key)
+    signature = Signature.new(keypair)
+    source_account = Account.new(public_key)
+    {:ok, seq_num} = Accounts.fetch_next_sequence_number(public_key)
+    sequence_number = SequenceNumber.new(seq_num)
```
Comment (juanhurtado10) [inline lib/contract/invoke_host_function.ex:54]: @Odraxs the use of a `with` clause will make your function response and error handling predictable.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/contract/invoke_host_function.ex [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    invoke_host_function_op
+    |> InvokeHostFunction.set_footprint(footprint)
+    |> InvokeHostFunction.set_contract_auth(authorization)
+  end
+
+  defp set_invoke_host_function_params(invoke_host_function_op, footprint, nil, _auth_account),
+    do: InvokeHostFunction.set_footprint(invoke_host_function_op, footprint)
+
+  @spec covert_to_sc_val(function_args :: function_args()) :: {:ok, sc_val_list()}
+  defp covert_to_sc_val(function_args) do
```
Comment (miguelnietoa) [inline lib/contract/invoke_host_function.ex:169]: Fix typo

File: lib/contract/contract.ex
Comment (miguelnietoa) [inline lib/contract/contract.ex:1]: This file should be placed in `lib/`

File: test/contract/contract_test.exs [hunk 4]
```diff
@@ -0,0 +1,136 @@
+defmodule Stellar.Horizon.Client.ContractCannedAccountRequests do
```
Comment (miguelnietoa) [inline test/contract/contract_test.exs:1]: Please improve this module name, it should always start with `Canned...`

File: test/contract/contract_test.exs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        @base_url <> "/accounts/GBNDWIM7DPYZJ2RLJ3IESXBIO4C2SVF6PWZXS3DLODJSBQWBMKY5U4M3",
+        _headers,
+        _body,
+        _opts
+      ) do
+    {:ok, 200, [], "{\"sequence\":\"1390916568875069\"}"}
+  end
+end
+
+defmodule Soroban.RPC.ContractCannedClientImpl do
```
Comment (miguelnietoa) [inline test/contract/contract_test.exs:17]: Same as above!

**Review by miguelnietoa — CHANGES_REQUESTED** [review]

File: lib/contract/invoke_host_function.ex [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    with {public_key, _secret} = keypair <- Stellar.KeyPair.from_secret_seed(source_secret_key),
+         {:ok, seq_num} <- Accounts.fetch_next_sequence_number(public_key),
+         {:ok, function_args} <- covert_to_sc_val(function_args),
+         signature <- Signature.new(keypair),
+         source_account <- Account.new(public_key),
+         sequence_number <- SequenceNumber.new(seq_num),
+         auth_account <- Enum.at(auth_accounts, 0) do
+      invoke_host_function_op = create_host_function_op(contract_id, function_name, function_args)
+
+      invoke_host_function_op
```
[the remark this answers is not in the document]
Comment (juanhurtado10) [reply lib/contract/invoke_host_function.ex:58]: @Odraxs I mean .... 👇 

```elixir
contract_id
|> create_host_function_op(function_name, function_args)
|> simulate(source_account, sequence_number)
|> send_transaction(...)
```

[the remark this answers is not in the document]
Comment (Odraxs) [reply lib/contract/invoke_host_function.ex:58]: @juanhurtado10 then, how do I send the result of `create_host_function_op` to `send_transaction`?

image

**Review by juanhurtado10 — APPROVED** [review]
LGTM 🚀

**Review by miguelnietoa — APPROVED** [review]
Well done, guys! 🚀

**Review by miguelnietoa — APPROVED** [review]

**Pull request merged**